### ISSUE TITLE
Retry on failing connection

### ISSUE THREAD

**Issue by AlexZeitler** [robconery/massive-js#93]
Does Massive support sort of retry pattern if the connection to Postgres fails?

I created a test script like this using two Docker containers (yes I could have linked them but that doesn't solve the issues :grin:)

````
docker run --name some-postgres -e POSTGRES_PASSWORD=postgres -p 5432:5432 -d postgres
docker run -it -d --name sf-customerservice -p 3000:3000 customerservice
````

When I run my tests, the `some-postgres` container has started but the database doesn't seem to be available at this time - result:

````
assert.js:86
  throw new assert.AssertionError({
        ^
AssertionError: Error: Connection terminated
    at /home/alex/src/hellomassive/customerservice/node_modules/massive/lib/runner.js:32:12
    at /home/alex/src/hellomassive/customerservice/node_modules/massive/node_modules/pg/lib/pool.js:75:25
    at /home/alex/src/hellomassive/customerservice/node_modules/massive/node_modules/pg/node_modules/generic-pool/lib/generic-pool.js:274:11
    at /home/alex/src/hellomassive/customerservice/node_modules/massive/node_modules/pg/lib/pool.js:27:26
    at null.<anonymous> (/home/alex/src/hellomassive/customerservice/node_modules/massive/node_modules/pg/lib/client.js:179:7)
    at g (events.js:199:16)
    at emit (events.js:104:17)
    at Socket.<anonymous> (/home/alex/src/hellomassive/customerservice/node_modules/massive/node_modules/pg/lib/connection.js:66:10)
    at Socket.emit (events.js:107:17)
    at TCP.close (net.js:485:12)
    at Function.module.exports.runLoopOnce (/home/alex/src/hellomassive/customerservice/node_modules/massive/node_modules/deasync/index.js:54:10)
    at Object.connectSync (/home/alex/src/hellomassive/customerservice/node_modules/massive/node_modules/deasync/index.js:31:19)
    at Object.<anonymous> (/home/alex/src/hellomassive/customerservice/app.js:14:31)
    at Module._compile (module.js:460:26)
    at Object.Module._extensions..js (module.js:478:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Module.require (module.js:365:17)
    at require (module.js:384:17)
    at Object.<anonymous> (/home/alex/src/hellomassive/customerservice/test/test.js:2:11)
    at Module._compile (module.js:460:26)
    at Object.Module._extensions..js (module.js:478:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Module.require (module.js:365:17)
    at require (module.js:384:17)
    at /home/alex/src/hellomassive/customerservice/node_modules/mocha/lib/mocha.js:192:27
    at Array.forEach (native)
    at Mocha.loadFiles (/home/alex/src/hellomassive/customerservice/node_modules/mocha/lib/mocha.js:189:14)
    at Mocha.run (/home/alex/src/hellomassive/customerservice/node_modules/mocha/lib/mocha.js:422:31)
    at Object.<anonymous> (/home/alex/src/hellomassive/customerservice/node_modules/mocha/bin/_mocha:398:16)
    at Module._compile (module.js:460:26)
    at Object.Module._extensions..js (module.js:478:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Function.Module.runMain (module.js:501:10)
    at startup (node.js:129:16)
    at node.js:814:3
````

If I place a `sleep 1` between the two container commands above, it works.
Anyway, temporary loss of connection is always possible even when already running... :smile:

**Comment by robconery** [comment]
I suppose we could try that - I'm a bit hesitant because failure to connect should be handled by you, not by Massive :)... however I'm all for convenience. 

If you want to PR this - go for it. Have a look in runner.js - that's where we connect. You can catch the connection error and be sure to console.log that an error connecting occurred, mentioning a single retry (and only a single retry) in 2 seconds or so.

Sound good?

**Comment by AlexZeitler** [comment]
Sounds like a plan.

**Comment by AlexZeitler** [comment]
Is it ok to add Docker for testing purposes (spin up a container, create the `rob:password@localhost/massive` database, run the tests)

**Comment by robconery** [comment]
Hmmm - I’m not  docker fan and I don’t want people to have to figure it out just to run our tests… so I’d rather not...

**Comment by AlexZeitler** [comment]
Ok. Do you have another idea how we could simulate a temporarily dropped connection for testing purposes?

**Comment by robconery** [comment]
Hmm - I don't... sorry...

**Comment by dmfay** [comment]
Might be possible to stub it with sinon to simulate the fault and then use stub.restore() to put the method back?

**Comment by AlexZeitler** [comment]
That might works, thanks.

**Comment by AlexZeitler** [comment]
Sorry for being late on this.

I tried to reconnect in the `runner.js` as suggested but that didn't work out as expected.

So I tried to handle the error in my application (as suggested also):

async:
````js
massive.connect({connectionString: connectionString, scriptsDir: '../db'}, function (err, db) {
    if (err) {
        console.log('connection err')
    }
});
````

This results in that output:
````bash
assert.js:86
  throw new assert.AssertionError({
        ^
AssertionError: Error: connect ECONNREFUSED
    at /app/node_modules/massive/lib/runner.js:32:12
    at /app/node_modules/massive/node_modules/pg/lib/pool.js:75:25
    at /app/node_modules/massive/node_modules/pg/node_modules/generic-pool/lib/generic-pool.js:274:11
    at /app/node_modules/massive/node_modules/pg/lib/pool.js:27:26
    at null.<anonymous> (/app/node_modules/massive/node_modules/pg/lib/client.js:171:5)
    at emit (events.js:107:17)
    at Socket.<anonymous> (/app/node_modules/massive/node_modules/pg/lib/connection.js:59:10)
    at Socket.emit (events.js:107:17)
    at net.js:459:14
    at process._tickCallback (node.js:355:11)

````

So I cannot handle `err`.

Then I tried to use the sync version using try / catch:

````js
try {
    var db = massive.connectSync({connectionString: connectionString, scriptsDir: '../db'});
} catch (e) {
    console.log('exception', e)
}
````

But it never hits the `catch` path:

````bash
assert.js:86
  throw new assert.AssertionError({
        ^
AssertionError: Error: connect ECONNREFUSED
    at /app/node_modules/massive/lib/runner.js:32:12
    at /app/node_modules/massive/node_modules/pg/lib/pool.js:75:25
    at /app/node_modules/massive/node_modules/pg/node_modules/generic-pool/lib/generic-pool.js:274:11
    at /app/node_modules/massive/node_modules/pg/lib/pool.js:27:26
    at null.<anonymous> (/app/node_modules/massive/node_modules/pg/lib/client.js:171:5)
    at emit (events.js:107:17)
    at Socket.<anonymous> (/app/node_modules/massive/node_modules/pg/lib/connection.js:59:10)
    at Socket.emit (events.js:107:17)
    at net.js:459:14
    at process._tickCallback (node.js:355:11)
````

**Comment by AlexZeitler** [comment]
This might be related: http://stackoverflow.com/questions/15969770/how-to-handle-should-js-assert-error

**Comment by AlexZeitler** [comment]
Using this, I can keep the application running but it feels wrong:

````js
process.on('uncaughtException', function (err) {
    console.error(err);
    console.log("Node NOT Exiting...");
});
````

I also tried it using plain `pg` which allows me to handle my errors and retry connecting.

**Comment by AlexZeitler** [comment]
I think it could be fixed by instead of throwing connection errors just bubbling them up and thus using existing retry implementations like [node-retry](https://github.com/tim-kos/node-retry).

**Comment by AlexZeitler** [comment]
@dmfay @robconery I would be very happy to hear your opinion about how to proceed. Thanks.

**Comment by robconery** [comment]
Sorry about that - have been traveling (again!). I'm tossing this around in my mind - I strongly dislike libraries throwing and halting the thread - but then again if a connection can't be made, that's a very haltable kind of thing. By bubbling the connection error you're handing the "can we go on" decision off... which I guess is precisely what you want...

Hmmm  - OK go for it :). Let's see what happens...

**Comment by AlexZeitler** [comment]
Ok, thanks!

**Comment by AlexZeitler** [comment]
Just wanted to make sure I broke nothing and tried to run the tests before changing anything.

I ran `mocha ./test/*.js` and got this result:

````bash
  Document queries
    ✓ returns a db
    Querying documents
      ✓ returns all documents when passed "*"
      ✓ returns all documents when passed only "next" function
      ✓ finds a doc by primary key
      ✓ finds a doc with > comparison on primary key
      ✓ finds a doc with >= comparison on primary key
      ✓ finds a doc by title
      ✓ parses greater than with two string defs
      ✓ parses greater than with a numeric
      ✓ parses less than with a numeric
      ✓ deals with arrays using IN
      ✓ deals with arrays using NOT IN
      ✓ executes a contains if passed an array of objects
      ✓ works properly with dates
    Full Text Search
      ✓ works on single key
      ✓ works on multiple key
      ✓ returns multiple results
      ✓ returns properly formatted documents with id etc

  Document saves
    To a non-existent table
      ✓ creates the table
      ✓ returns the doc
      ✓ updates the doc
      ✓ finds the updated document
      ✓ deletes the doc
    To an Existing Table
      ✓ Saves a new movie
      ✓ updates the movie title
      ✓ finds the updated movie title
      ✓ deletes the movie

  Functions
    ✓ executes all products
    ✓ executes all myschema.albums
    ✓ executes artists with param
    Loading of Functions from PG
      ✓ has an all_products function attached
      ✓ has a schema-bound function attached to myschema
    Function Execution
      ✓ executes all_products and returns the results
      ✓ executes schema-bound function and returns the results
    Functions with Cased Names
      ✓ executes camel-cased function AllMyProducts and returns the results
      ✓ executes schema-bound, camel-cased function AllMyAlbums and returns the results

  On spin up
    ✓ returns a valid db interface
    ✓ loads non-public schema as namespace property
    ✓ loads tables in db schema as properties of namespace
    ✓ loads up 7 tables with 2 in schema object in array property
    ✓ loads up 7 queries
    1) loads up functions

  Synchronous Load
    2) loads the db synchronously and blocks execution until complete
    3) returns a valid db instance from sync load function
    4) loads non-public schema as namespace property
    5) loads tables in syncLoaded schema as properties of namespace
    6) loads up 7 tables with 2 in schema object in array property
    7) loads up 7 queries
    8) loads up functions

  On Load with Schema Filters (these tests may run slow - loads db each test!!)
    9) loads all schema tables when no schema argument is passed
    10) loads all schema tables when passed "all" as schema argument
    11) loads only tables from specified schema when passed schema name as schema argument
    12) loads only tables from public schema when passed "public" name as schema argument
    13) loads only tables from specified schema when passed comma-delimited list of schema names
    14) loads only tables from specified schema when passed an array of schema names

  On Load with Table blacklist (these tests may run slow - loads db each test!!)
    15) loads all tables when no blacklist argument is provided
    16) excludes tables with name matching blacklist pattern as a string argument
    17) excludes tables with name and schema matching blacklist pattern as a string argument
    18) excludes tables with name and schema matching multiiple blacklist patterns as a comma-delimited string argument
    19) excludes tables with name and schema matching multiiple blacklist patterns as a string array argument
    20) allows exceptions to schema filter
    21) allows exceptions to blacklist filter
    22) allows exceptions to schema and blacklist filters

  On Load with Table whitelist (these tests may run slow - loads db each test!!)
    23) loads all tables when no whitelist argument is provided
    24) includes ONLY tables with name matching whitelisted table names as a string argument
    25) includes ONLY tables with name matching whitelisted items in comma-delimited string
    26) includes ONLY tables with name matching whitelisted items in string array
    27) whitelist overrides other filters

  On load with Function Exclusion (these tests may run slow - loads db each test!!)
    28) excludes functions at load whenever it is told...
    29) includes all functions at load whenever it is told...
    30) includes all functions at load by default...

  On load with Function Blacklist (these tests may run slow - loads db each test!!)
    31) loads all functions when no blacklist argument is provided
    32) excludes functions with name matching blacklist pattern as a string argument
    33) excludes functions with name and schema matching blacklist pattern as a string argument
    34) excludes functions with name and schema matching multiiple blacklist patterns as a comma-delimited string argument
    35) excludes functions with name and schema matching multiiple blacklist patterns as a string array argument

  Queries built from files
    Loading of queries
      ✓ returns a db
      ✓ has a schema query attached
      ✓ has an inStockProducts query attached
    Execution
      ✓ executes inStockProducts without args with only a callback
      ✓ executes productById with non-array arg and callback
      ✓ executes productByName with multiple args and callback
      ✓ executes a deep namespaced query

  Schmema-Bounds Document queries
    ✓ returns a db
    Querying documents in a schema-bound table
      ✓ returns all documents when passed "*"
      ✓ returns all documents when passed only "next" function
      ✓ finds a doc by primary key
      ✓ finds a doc with > comparison on primary key
      ✓ finds a doc with >= comparison on primary key
      ✓ finds a doc by title
      ✓ parses greater than with two string defs
      ✓ parses greater than with a numeric
      ✓ parses less than with a numeric
      ✓ deals with arrays using IN
      ✓ deals with arrays using NOT IN
      ✓ executes a contains if passed an array of objects
      ✓ works properly with dates
    Full Text Search in a schema-bound document table
      ✓ works on single key
      ✓ works on multiple key
      ✓ returns multiple results
      ✓ returns properly formatted documents with id etc

  Schema-Bound Document Saves
    To a non-existent table
      ✓ creates the table
      ✓ returns the doc
      ✓ updates the doc
      ✓ finds the updated document
      ✓ deletes the doc
    To an Existing Table
      ✓ Saves a new movie
      ✓ updates the movie title
      ✓ finds the updated movie title
      ✓ deletes the movie

  Schema - Bound Tables -Add/Edit/Delete
    Add/Update/Delete records in a schema-bound table:
      ✓ adds an artist 
      ✓ updates an artist 
      ✓ deletes an artist 

  Schema - Bound Tables -Querying
    Simple queries with args
      ✓ returns Artist 1 with 1 as only arg
      ✓ returns first record with findOne no args
    Simple queries without args
      ✓ returns all records on find with no args
      ✓ returns first record with findOne no args
    Simple queries with AND and OR
      ✓ returns Artist 1 OR Artist 2
      ✓ returns Artist 1 AND Artist 2
      ✓ returns Artist 1 with params as not array
    Simple queries with count
      ✓ returns 2 for OR id 1 or 2
      ✓ returns 1 for id 1
    Simple comparative queries
      ✓ returns Artist with id greater than 2
      ✓ returns Artist with id less than 2
      ✓ returns myschema.artists IN 1 and 2
      ✓ returns Artist NOT IN 1 and 2
    Limiting and Offsetting results
      ✓ returns 1 result with limit of 1
      ✓ returns second result with limit of 1, offset of 1
      ✓ returns id and name if sending in columns
    Ordering results
      ✓ returns ascending order of myschema.artists by name
      ✓ returns descending order of myschema.artists

  Synchronous goodies
    36) "before all" hook

  96 passing (2s)
  36 failing

  1) On spin up loads up functions:

      AssertionError: 30 == 20
      + expected - actual

      -30
      +20
      
      at Context.<anonymous> (test/loader_spec.js:31:12)

/home/alex/.nvm/versions/node/v0.12.3/lib/node_modules/mocha/lib/reporters/base.js:199
      var match = message.match(/^([^:]+): expected/);
                          ^
TypeError: undefined is not a function
    at /home/alex/.nvm/versions/node/v0.12.3/lib/node_modules/mocha/lib/reporters/base.js:199:27
    at Array.forEach (native)
    at Function.exports.list (/home/alex/.nvm/versions/node/v0.12.3/lib/node_modules/mocha/lib/reporters/base.js:161:12)
    at Spec.Base.epilogue (/home/alex/.nvm/versions/node/v0.12.3/lib/node_modules/mocha/lib/reporters/base.js:319:10)
    at Runner.emit (events.js:129:20)
    at Runner.uncaught (/home/alex/.nvm/versions/node/v0.12.3/lib/node_modules/mocha/lib/runner.js:593:8)
    at process.uncaught (/home/alex/.nvm/versions/node/v0.12.3/lib/node_modules/mocha/lib/runner.js:612:10)
    at process.emit (events.js:107:17)
    at process._fatalException (node.js:236:26)
    at Function.module.exports.runLoopOnce (/home/alex/src/massive-js/node_modules/deasync/index.js:54:10)
    at Object.connectSync (/home/alex/src/massive-js/node_modules/deasync/index.js:31:19)
    at Context.<anonymous> (/home/alex/src/massive-js/test/sync_spec.js:8:18)
    at callFn (/home/alex/.nvm/versions/node/v0.12.3/lib/node_modules/mocha/lib/runnable.js:266:21)
    at Hook.Runnable.run (/home/alex/.nvm/versions/node/v0.12.3/lib/node_modules/mocha/lib/runnable.js:259:7)
    at next (/home/alex/.nvm/versions/node/v0.12.3/lib/node_modules/mocha/lib/runner.js:271:10)
    at Immediate._onImmediate (/home/alex/.nvm/versions/node/v0.12.3/lib/node_modules/mocha/lib/runner.js:292:5)
    at processImmediate [as _immediateCallback] (timers.js:358:17)
    at Function.module.exports.runLoopOnce (/home/alex/src/massive-js/node_modules/deasync/index.js:54:10)
    at Object.loadSync (/home/alex/src/massive-js/node_modules/deasync/index.js:31:19)
    at Context.<anonymous> (/home/alex/src/massive-js/test/loader_spec.js:44:38)
    at callFn (/home/alex/.nvm/versions/node/v0.12.3/lib/node_modules/mocha/lib/runnable.js:266:21)
    at Test.Runnable.run (/home/alex/.nvm/versions/node/v0.12.3/lib/node_modules/mocha/lib/runnable.js:259:7)
    at Runner.runTest (/home/alex/.nvm/versions/node/v0.12.3/lib/node_modules/mocha/lib/runner.js:390:10)
    at /home/alex/.nvm/versions/node/v0.12.3/lib/node_modules/mocha/lib/runner.js:473:12
    at next (/home/alex/.nvm/versions/node/v0.12.3/lib/node_modules/mocha/lib/runner.js:315:14)
    at /home/alex/.nvm/versions/node/v0.12.3/lib/node_modules/mocha/lib/runner.js:325:7
    at next (/home/alex/.nvm/versions/node/v0.12.3/lib/node_modules/mocha/lib/runner.js:260:23)
    at Immediate._onImmediate (/home/alex/.nvm/versions/node/v0.12.3/lib/node_modules/mocha/lib/runner.js:292:5)
    at processImmediate [as _immediateCallback] (timers.js:358:17)
````

Maybe I did not set up the expected database correct - is there a script or do some tests really fail currently?

**Comment by dmfay** [comment]
There's no setup script but it's pretty simple -- make sure you have a user rob/password (you may need to change your postgres config to require password authentication locally, I can't remember offhand) with full rights on a database named 'massive'. The connection string is in test/helpers/index.js. The tests bootstrap all the schema & data they need, and I'm not seeing any failures at the moment.

**Comment by xivSolutions** [comment]
Have you run `npm install` recently? Some of the recent changes you may have pulled down involved new libraries which may cause code to break if they are not in place. 

Also note - in the test/db/schema.sql file is where you will find the "setup" script, such as it is. the test/helpers/index.js file performs the basic setup in advance of each test module.

Also, review the actual error report, and see what may be going on. For example, the first case, it appears you have more functions loading than there are in the existing code. The `assert` shows you that there are 30 functions found, against a test condition of 20 (see the loader_spec.js test file - the test expects 20 functions, and according to your mocha output, 30 are being reported). 

In the current incarnation of massive, as pulled down directly from the master repo, all tests should be passing.

**Comment by AlexZeitler** [comment]
After re-installing Postgres 9.4 the tests ran - but I had to run them twice - first run failed.

**Comment by ldesplat** [comment]
I really like this. I try very hard to keep away from libraries that throw. You can catch the exception by using `domains`. Since it is async underneath, try/catch will never work. BTW, domains have a lot of issues but it's what we have.

```
var Domain = require('domain');
var d = Domain.create();

d.on('error', function (error) {

  console.log(error);
});

d.run(function () {

  massive.connect({connectionString: connectionString, scriptsDir: '../db'}, function (err, db) {

    if (err) {
      console.log('connection err')
    }
  });
});
```

But then, I agree that if it throws only during connect then I can understand keeping it. Still, if there is an error bubbled up during connect.. I would handle it myself quite easily and most likely prevent the application from starting up.

**Comment by AlexZeitler** [comment]
@ldesplat Thanks, using the `domains` approach catching the error works. 
I have already created a PR for bubbling: https://github.com/robconery/massive-js/pull/104

**Comment by robconery** [comment]
I've been traveling again; I'll tackle this next week...

**Comment by AlexZeitler** [comment]
No worries, take your time.

**Comment by robconery** [comment]
OK I'm bubbling the connection error so you should be able to handle (hopefully). Published 2.0.8 just now...

**Comment by AlexZeitler** [comment]
Thanks, but error handling now is broken if no connection can be made:
```
/home/alex/code/massive-retry/node_modules/massive/lib/runner.js:36
      next(err,null);
      ^
ReferenceError: next is not defined
```

I've send you an updated [PR](https://github.com/robconery/massive-js/pull/120) that fixes `2.0.8` and works with retry. (`index.js` also needs to be updated to work with retry)
